Install includes to include/${PROJECT_NAME} and use modern CMake

sloretz opened the pull request:
Part of ros2/ros2#1150

This installs includes to `include/${PROJECT_NAME}` to mitigate include directory search order issues when overriding packages in desktop.

Part of ament/ament_cmake#365

This removes `ament_export_libraries` and `ament_export_include_directories` as they're redundant with the exported CMake targets in packages that I was already changing. I only updated packages that installed headers.

Part of ament/ament_cmake#292

This replaces `ament_target_dependencies()` calls with `target_link_libraries()` in packages that I was already changing. I only updated packages that installed headers.

This also removes the dependency on `eigen3_cmake_module` in packages that I was already changing. It's no longer necessary now that we can use modern CMake targets. I only updated packages that installed headers.

Requires ros2/rclcpp#1855 for the `rclcpp_components::component` target.

sloretz committed:
Install includes to include/${PROJECT_NAME} and use modern CMake

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove dependency on eigen3_cmake_module

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

clalancette commented on test_tf2/CMakeLists.txt:
```diff
@@ -36,92 +35,83 @@ ament_find_gtest()
-  )
-  target_link_libraries(test_buffer_client ${GTEST_LIBRARIES})
+  target_link_libraries(test_buffer_client
+    ${GTEST_LIBRARIES}
```
Just a question: is there no modern CMake target for GTest?

sloretz replied:
Their might be. I didn't look :no_mouth:. If so it's probably in the `GTEST_LIBRARIES` variable.

sloretz replied:
I suspect the `${GTEST_LIBRARIES}` variable always contains the modern CMake target `gtest`, but I haven't dug through `ament_cmake_gtest` enough to be confident in that. I'll leave this as is for now.

clalancette commented on tf2/CMakeLists.txt:
```diff
@@ -92,9 +90,6 @@ if(BUILD_TESTING)
   ament_add_gtest(test_simple_tf2_core test/simple_tf2_core.cpp)
   if(TARGET test_simple_tf2_core)
     target_link_libraries(test_simple_tf2_core tf2)
-    ament_target_dependencies(test_simple_tf2_core
-      "geometry_msgs"
-    )
```
This test does have a dependency on `geometry_msgs`, so shouldn't this change to a `target_link_libraries`?

sloretz replied:
https://github.com/ros2/geometry2/pull/493#discussion_r840113294

clalancette commented on tf2/CMakeLists.txt:
```diff
@@ -104,7 +99,6 @@ if(BUILD_TESTING)
 endif()
 
 ament_export_dependencies(console_bridge geometry_msgs rcutils rosidl_runtime_cpp)
-ament_export_include_directories(include)
-ament_export_libraries(tf2)
```
Is this going to break downstream consumers that rely on the old CMake variables?

sloretz replied:
It would. I added back all the old-style CMake variable stuff in https://github.com/ros2/geometry2/pull/493/commits/92332fb40508dd6dde837b7ba6b9d1b2a4143d96

clalancette commented on tf2_eigen/CMakeLists.txt:
```diff
@@ -14,11 +14,20 @@ find_package(geometry_msgs REQUIRED)
+  "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+  "$<INSTALL_INTERFACE:include/${PROJECT_NAME}>")
+
+
```
Nit: extra newline.

sloretz replied:
Removed extra newline in https://github.com/ros2/geometry2/pull/493/commits/182554734478d9fa8904ca45dfb78a90c836ea06

clalancette commented on tf2_eigen/CMakeLists.txt:
```diff
@@ -35,20 +44,14 @@ if(BUILD_TESTING)
-      "tf2_ros")
-    target_include_directories(tf2_eigen-test PUBLIC
-      ${Eigen3_INCLUDE_DIRS})
+    target_link_libraries(tf2_eigen-test tf2_eigen)
```
It looks like this test depends on geometry_msgs, tf2, tf2_eigen, and Eigen.  Shouldn't we be listing all of them in this `target_link_libraries` call?

sloretz replied:
In https://github.com/ros2/geometry2/pull/493/commits/7dac8dd51997b0d0892a0e527e12c5f738b98d24 , https://github.com/ros2/geometry2/pull/493/commits/2a9df115fcde75dbd54ddc9347e55d66c1dbedd7 , and https://github.com/ros2/geometry2/pull/493/commits/e78a34d4724545ce557b9cf568d66819dfcd6c6d I double checked the dependencies, and applied the method we came up with in #496 of documenting some dependencies of tests instead of linking them.

clalancette commented on tf2_kdl/CMakeLists.txt:
```diff
@@ -35,18 +46,12 @@ if(BUILD_TESTING)
-    tf2
-    tf2_ros
-    tf2_msgs)
+  target_link_libraries(test_kdl
+    tf2_kdl
+    rclcpp::rclcpp
+    ${tf2_msgs_TARGETS})
```
This test also depends on tf2, tf2_ros, and kdl.  Shouldn't we depend on them here?

sloretz replied:
In https://github.com/ros2/geometry2/pull/493/commits/7dac8dd51997b0d0892a0e527e12c5f738b98d24 , https://github.com/ros2/geometry2/pull/493/commits/2a9df115fcde75dbd54ddc9347e55d66c1dbedd7 , and https://github.com/ros2/geometry2/pull/493/commits/e78a34d4724545ce557b9cf568d66819dfcd6c6d I double checked the dependencies, and applied the method we came up with in #496 of documenting some dependencies of tests instead of linking them.

clalancette commented on tf2_ros/CMakeLists.txt:
```diff
@@ -82,45 +67,31 @@ target_link_libraries(static_transform_publisher
   ${PROJECT_NAME}
   static_transform_broadcaster_node
 )
```
This executable also depends on tf2 and rclcpp; shouldn't it have dependencies on those?

sloretz replied:
I fixed this executable's dependencies in https://github.com/ros2/geometry2/pull/493/commits/e78a34d4724545ce557b9cf568d66819dfcd6c6d

clalancette reviewed (commented):
I've left a number of questions inline.  I basically have two overarching questions:

1. If we remove `ament_export_include_directories` and `ament_export_libraries`, what happens to downstream projects that are depending on the old-style CMake variables?
2. I'm fine with switching from `ament_target_dependencies` to `target_link_libraries`  where we can, but shouldn't we be explicitly listing all dependencies in `target_link_libraries`?  There seem to be a number of places where we are now implicitly depending on other targets.

sloretz commented:
> If we remove ament_export_include_directories and ament_export_libraries, what happens to downstream projects that are depending on the old-style CMake variables?

So far, they all still work. The reason is all the downstream projects I've come across are using `ament_target_dependencies()`, which reads either old-style CMake variables or CMake targets.

A common CMake pattern I've seen outside of ROS is to put exported targets into a "..._LIBRARIES" variable taking advantage of downstream targets are likely already doing `target_link_libraries(...)` with it. However, `ament_cmake` puts the targets into a separate `..._TARGETS` variable, so that doesn't apply here.


> I'm fine with switching from ament_target_dependencies to target_link_libraries where we can, but shouldn't we be explicitly listing all dependencies in target_link_libraries? There seem to be a number of places where we are now implicitly depending on other targets.

I've been intentionally leaving it off when a target is getting dependencies from another target in the same CMake project. My reasoning is it tests that the CMake target we're exporting properly made those targets as `PUBLIC` or `INTERFACE`. Say there was a package with a shared library and a test executable. Now say there's a bug where the shared library says it depends on `rclcpp::rclcpp` with the `PRIVATE` keyword, but it should have used the `PUBLIC` keyword. If the test also `target_link_libraries(... rclcpp::rclcpp)` then the project will build fine and the tests will pass, but when that package is released downstream projects will fail to build because the shared library target didn't properly export it's dependency on `rclcpp::rclcpp`. If the test instead got that dependency from the shared library target then the test would have failed to build and caught that bug before release.

sloretz committed:
Undo whitespace change

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove extra newline

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Remove duplicate add_library call

sloretz committed:
Keep making old style CMake variables

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Update test dependencies

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Alphabetize

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Include dirs and libs for static_transform_broadcaster_node

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
I rebased this and addressed all feedback.

CI
* build: `--packages-above-and-dependencies examples_tf2_py geometry2 test_tf2 tf2 tf2_bullet tf2_eigen tf2_eigen_kdl tf2_geometry_msgs tf2_kdl tf2_msgs tf2_py tf2_ros tf2_ros_py tf2_sensor_msgs tf2_tools`
* test: `--packages-select examples_tf2_py geometry2 test_tf2 tf2 tf2_bullet tf2_eigen tf2_eigen_kdl tf2_geometry_msgs tf2_kdl tf2_msgs tf2_py tf2_ros tf2_ros_py tf2_sensor_msgs tf2_tools`

Jobs
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=16495)](http://ci.ros2.org/job/ci_linux/16495/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11084)](http://ci.ros2.org/job/ci_linux-aarch64/11084/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16868)](http://ci.ros2.org/job/ci_windows/16868/)

sloretz commented:
> https://github.com/ros2/choco-packages/blob/latest/package/eigen/share/cmake/Eigen3Config.cmake

Windows CI failed because we ship a custom Eigen3Config.cmake on Windows that doesn't provide targets instead of using upstreams. This is the same reason I upgraded bullet, but I don't have time to upgrade that one before the freeze. I'll undo all the target stuff for packages using Eigen to get Windows to pass.

sloretz committed:
Use Eigen3::Eigen when available, else Eigen3_INCLUDE_DIRS

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
5cedb80 should resolve the windows issue. CI rerun:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=16497)](http://ci.ros2.org/job/ci_linux/16497/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11086)](http://ci.ros2.org/job/ci_linux-aarch64/11086/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16870)](http://ci.ros2.org/job/ci_windows/16870/)

clalancette commented on test_tf2/CMakeLists.txt:
```diff
@@ -36,92 +35,88 @@ ament_find_gtest()
+      # TODO(sloretz) require target to exist when https://github.com/ros2/choco-packages/issues/19 is addressed
+      target_link_libraries(test_convert Eigen3::Eigen)
+    else()
+      target_include_directories(test_convert ${Eigen3_INCLUDE_DIRS})
```
Windows CI is complaining; I think you need `PRIVATE` here.

sloretz replied:
Added `PRIVATE` in 06a70c708efcb48283985ef02d3aa5fe36a542ef

clalancette commented on test_tf2/CMakeLists.txt:
```diff
@@ -36,92 +35,88 @@ ament_find_gtest()
+    tf2_eigen::tf2_eigen
+    ${tf2_geometry_msgs_TARGETS}
+    tf2_kdl::tf2_kdl)
+    if(TARGET Eigen3::Eigen)
```
The indentation on this block seems weird.  I think it should be two spaces to the left, otherwise it looks like an argument to `target_link_libraries` above.

sloretz replied:
Fixed indentation in 255059df33b1c09caabf651430c7c0a1d33a9b2f

clalancette commented on tf2_ros/CMakeLists.txt:
```diff
@@ -80,47 +73,38 @@ add_executable(static_transform_publisher
   static_transform_broadcaster_node
-)
-ament_target_dependencies(static_transform_publisher
-  "geometry_msgs"
-  "rclcpp"
-  "tf2"
-  "tf2_msgs"
```
`static_transform_publisher` depends on both `geometry_msgs` and `tf2_msgs` directly, so I think we should keep them as dependencies.  Or, if we are testing out transitive dependencies, at least have a comment.  The same goes for the two targets below.

sloretz replied:
Are you looking at `static_transform_broadacster_node`? I don't see either `geometry_msgs` or `tf2_msgs` used by [`static_transform_publisher`](https://github.com/ros2/geometry2/blob/sloretz__easy_idso_part1/tf2_ros/src/static_transform_broadcaster_program.cpp). Their source files have confusingly similar names (`static_transform_broadcaster_node.cpp` vs `static_transform_broadcaster_program.cpp`)

clalancette replied:
Oh, you are totally right.  I got confused by the target name vs. the filename.  Never mind this comment, this looks good.

clalancette reviewed (changes requested):
I've left a couple of minor things (and Windows CI needs to be fixed).  With those things done, I'll be happy to approve.

sloretz committed:
target_include_directories(... PRIVATE

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Fix indentation

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
CI re-run:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=16505)](http://ci.ros2.org/job/ci_linux/16505/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11092)](http://ci.ros2.org/job/ci_linux-aarch64/11092/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=16876)](http://ci.ros2.org/job/ci_windows/16876/)

sloretz commented:
CI LGTM :tada: @clalancette Mind taking another look?
